File: src/QxRibbonDrawHelper.cpp
/**
 * Copyleft (C) 2023 maminjie <canpool@163.com>
 * SPDX-License-Identifier: MIT
**/
#include "QxRibbonDrawHelper.h"

RibbonDrawHelper::RibbonDrawHelper()
{
}

QPixmap RibbonDrawHelper::iconToPixmap(const QIcon &icon, QWidget *widget, const QStyleOption *opt,
                                       const QSize &icoSize)
{
    QIcon::Mode mode;
    QIcon::State state;

    if (!(opt->state & QStyle::State_Enabled)) {
        mode = QIcon::Disabled;
    } else if ((opt->state & QStyle::State_MouseOver) && (opt->state & QStyle::State_AutoRaise)) {
        mode = QIcon::Active;
    } else {
        mode = QIcon::Normal;
    }

    if (opt->state & QStyle::State_Selected || opt->state & QStyle::State_On) {
        state = QIcon::On;
    } else {
        state = QIcon::Off;
    }
    return (icon.pixmap(widget->window()->windowHandle(), icoSize, mode, state));
}

void RibbonDrawHelper::drawIcon(const QIcon &icon, QPainter *painter, const QStyleOption *opt, int x, int y, int width,
                                int height)
{
    QIcon::Mode mode;
    QIcon::State state;

    if (!(opt->state & QStyle::State_Enabled)) {
        mode = QIcon::Disabled;
    } else if ((opt->state & QStyle::State_MouseOver) && (opt->state & QStyle::State_AutoRaise)) {
        mode = QIcon::Active;
    } else {
        mode = QIcon::Normal;
    }

    if (opt->state & QStyle::State_Selected || opt->state & QStyle::State_On) {
        state = QIcon::On;
    } else {
        state = QIcon::Off;
    }

    icon.paint(painter, x, y, width, height, Qt::AlignCenter, mode, state);
}

void RibbonDrawHelper::drawIcon(const QIcon &icon, QPainter *painter, const QStyleOption *opt, const QRect &rect)
{
    QIcon::Mode mode;
    QIcon::State state;

    if (!(opt->state & QStyle::State_Enabled)) {
        mode = QIcon::Disabled;
    } else if ((opt->state & QStyle::State_MouseOver) && (opt->state & QStyle::State_AutoRaise)) {
        mode = QIcon::Active;
    } else {
        mode = QIcon::Normal;
    }

    if (opt->state & QStyle::State_Selected || opt->state & QStyle::State_On) {
        state = QIcon::On;
    } else {
        state = QIcon::Off;
    }

    icon.paint(painter, rect, Qt::AlignCenter, mode, state);
}

QSize RibbonDrawHelper::iconActualSize(const QIcon &icon, const QStyleOption *opt, const QSize &iconSize)
{
    QIcon::Mode mode;
    QIcon::State state;

    if (!(opt->state & QStyle::State_Enabled)) {
        mode = QIcon::Disabled;
    } else if ((opt->state & QStyle::State_MouseOver) && (opt->state & QStyle::State_AutoRaise)) {
        mode = QIcon::Active;
    } else {
        mode = QIcon::Normal;
    }

    if (opt->state & QStyle::State_Selected || opt->state & QStyle::State_On) {
        state = QIcon::On;
    } else {
        state = QIcon::Off;
    }
    return (icon.actualSize(iconSize, mode, state));
}

void RibbonDrawHelper::drawText(const QString &text, QStylePainter *painter, const QStyleOption *opt, Qt::Alignment al,
                                int x, int y, int width, int height)
{
    painter->drawItemText(QRect(x, y, width, height), al, opt->palette, opt->state & QStyle::State_Enabled, text);
}

void RibbonDrawHelper::drawText(const QString &text, QStylePainter *painter, const QStyleOption *opt, Qt::Alignment al,
                                const QRect &rect)
{
    painter->drawItemText(rect, al, opt->palette, opt->state & QStyle::State_Enabled, text);
}

File: pyqt/PyRibbon/QxRibbonDrawHelper.py
#!/usr/bin/env python3
# Copyright (c) 2023 maminjie <canpool@163.com>
# SPDX-License-Identifier: GPL-3.0

from PyQt5.QtCore import Qt, QSize, QRect
from PyQt5.QtGui import QPixmap, QIcon, QPainter
from PyQt5.QtWidgets import QWidget, QStyleOption, QStyle


class RibbonDrawHelper:

    @staticmethod
    def iconToPixmap(icon: QIcon, widget: QWidget, opt: QStyleOption, icon_size: QSize) -> QPixmap:
        mode = QIcon.Normal
        if not (opt.state & QStyle.State_Enabled):
            mode = QIcon.Disabled
        elif (opt.state & QStyle.State_MouseOver) and (opt.state & QStyle.State_AutoRaise):
            mode = QIcon.Active

        state = QIcon.Off
        if (opt.state & QStyle.State_Selected) or (opt.state & QStyle.State_On):
            state = QIcon.On

        return icon.pixmap(widget.window().windowHandle(), icon_size, mode, state)

    @staticmethod
    def drawIcon(icon: QIcon, painter: QPainter, opt: QStyleOption, *args):
        """
        drawIcon(QIcon, QPainter, QStyleOption, x: int, y: int, width: int, height: int)
        drawIcon(QIcon, QPainter, QStyleOption, rect: QRect)
        """
        if len(args) < 1:
            return
        if isinstance(args[0], int):
            rect = QRect(args[0], args[1], args[2], args[3])
        else:
            rect = args[0]

        mode = QIcon.Normal
        if not (opt.state & QStyle.State_Enabled):
            mode = QIcon.Disabled
        elif (opt.state & QStyle.State_MouseOver) and (opt.state & QStyle.State_AutoRaise):
            mode = QIcon.Active

        state = QIcon.Off
        if (opt.state & QStyle.State_Selected) or (opt.state & QStyle.State_On):
            state = QIcon.On

        icon.paint(painter, rect, Qt.AlignCenter, mode, state)

    @staticmethod
    def iconActualSize(icon: QIcon, opt: QStyleOption, icon_size: QSize) -> QSize:
        mode = QIcon.Normal
        if not (opt.state & QStyle.State_Enabled):
            mode = QIcon.Disabled
        elif (opt.state & QStyle.State_MouseOver) and (opt.state & QStyle.State_AutoRaise):
            mode = QIcon.Active

        state = QIcon.Off
        if (opt.state & QStyle.State_Selected) or (opt.state & QStyle.State_On):
            state = QIcon.On

        return icon.actualSize(icon_size, mode, state)

    @staticmethod
    def drawText(text: str, painter: QPainter, opt: QStyleOption, align, *args):
        """
        drawText(str, QPainter, QStyleOption, Qt.Alignment, x: int, y: int, width: int, height: int)
        drawText(str, QPainter, QStyleOption, Qt.Alignment, rect: QRect)
        """
        if len(args) < 1:
            return
        if isinstance(args[0], int):
            rect = QRect(args[0], args[1], args[2], args[3])
        else:
            rect = args[0]
        painter.drawItemText(rect, align, opt.palette, opt.state & QStyle.State_Enabled, text)
